src/gnosis-safe-multisig/agent.js

```diff
@@ -0,0 +1,213 @@
+	}));
+
+	return agentState;
+};
+
+const handleTransaction = async (agentState, txEvent) => {
+	const findings = [];
+
+	// if any transfers occurred to or from this safe, store the token address and create an ethers
+	// contract object for interactions in the handleBlock function
+	const transferLogs = txEvent.filterLog(agentState.transferSignature);
+	agentState.contracts.forEach((contract) => {
+		transferLogs.forEach((log) => {
+			const addressLower = contract.address.toLowerCase();
+			if (log.args.from.toLowerCase() !== addressLower && log.args.to.toLowerCase() !== addressLower) {
+				return;
+			}
+
+			const logAddressLower = log.address.toLowerCase();
+			const tokenAddresses = contract.tokenAddresses;
+			if ((tokenAddresses.indexOf(logAddressLower) === -1) && tokenAddresses.push(logAddressLower)) {
+				const tokenContract = new ethers.Contract(log.address, agentState.erc20Abi, agentState.provider);
+				contract.tokenContracts.push(tokenContract);
+			}
+		});
+
+		// filter for any events emitted by the safe contract
+		const logs = txEvent.filterLog(contract.eventSignatures, contract.address);
+		logs.forEach((log) => {
+			const findingObject = versionUtils.getFindings(contract.version, log.name, agentState.alertFields, contract.address, log.args);
+			if (!findingObject) {
+				return;
+			}
+
+			findingObject.type = FindingType.Info;
+			findingObject.severity = FindingSeverity.Info;
+			findingObject.protocol = agentState.alertFields.protocolName;
+			const finding = Finding.fromObject(findingObject);
+			findings.push(finding);
+		});
+	});
+
+	return findings;
+};
+
+const handleBlock = async (agentState) => {
+	const {
+		developerAbbreviation,
+		protocolAbbreviation,
+		protocolName,
+	} = agentState.alertFields;
+
+	// find changes in eth balance and tokens for every gnosis safe address
+	let totalFindings = await Promise.all(agentState.contracts.map(async (contract) => {
+		const { address } = contract;
+		const ethBalance = await agentState.provider.getBalance(address);
+		const ethBalanceBN = new BigNumber(ethBalance.toString());
+
+		// created new token contract for every token transfer that's ever happened to this address
+		const promises = contract.tokenContracts.map(async (tokenContract) => {
+			const result = {};
+			try {
+				// get the balance of each token for this specific safe
+				const tokenBalance = await tokenContract.balanceOf(address);
+				result[tokenContract.address.toLowerCase()] = new BigNumber(tokenBalance.toString());
+			} catch {
+				result[tokenContract.address.toLowerCase()] = new BigNumber(0);
+			}
+			return result;
+		});
+
+		const tokenBalances = {};
+		const tokenBalancesArray = await Promise.all(promises);
+		tokenBalancesArray.forEach((entry) => {
+			Object.entries(entry).forEach(([key, value]) => {
+				tokenBalances[key] = value;
+			});
+		});
+
+		const hasBalances = Object.keys(contract.previousBalances).length === 0;
+		if (!hasBalances) {
+			return [];
+		}
```

Comment:
In collaborative code review, we determined that this check does not need to occur. Line 162 should already take care of checking for the presence of keys.  If no keys exist, not iteration will occur.
```suggestion
```